From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 28 Mar 2016 20:26:34 -0400
Subject: [PATCH] Entity AddTo/RemoveFrom World Events


diff --git a/src/main/java/com/destroystokyo/paper/event/entity/EntityAddToWorldEvent.java b/src/main/java/com/destroystokyo/paper/event/entity/EntityAddToWorldEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..1d8e3c93a139bba11affca74b742269f24300d2c
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/event/entity/EntityAddToWorldEvent.java
@@ -0,0 +1,45 @@
+package com.destroystokyo.paper.event.entity;
+
+import org.bukkit.World;
+import org.bukkit.entity.Entity;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.EntityEvent;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Fired any time an entity is being added to the world for any reason (including a chunk loading).
+ * <p>
+ * Not to be confused with {@link CreatureSpawnEvent}
+ */
+public class EntityAddToWorldEvent extends EntityEvent {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+
+    @NotNull private final World world;
+
+    @ApiStatus.Internal
+    public EntityAddToWorldEvent(@NotNull Entity entity, @NotNull World world) {
+        super(entity);
+        this.world = world;
+    }
+
+    /**
+     * @return The world that the entity is being added to
+     */
+    @NotNull
+    public World getWorld() {
+        return this.world;
+    }
+
+    @NotNull
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+}
diff --git a/src/main/java/com/destroystokyo/paper/event/entity/EntityRemoveFromWorldEvent.java b/src/main/java/com/destroystokyo/paper/event/entity/EntityRemoveFromWorldEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..2f401ceb7477c416c71a3bb4c106de1951503b9a
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/event/entity/EntityRemoveFromWorldEvent.java
@@ -0,0 +1,126 @@
+package com.destroystokyo.paper.event.entity;
+
+import org.bukkit.World;
+import org.bukkit.entity.Entity;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.entity.EntityEvent;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Fired any time an entity is being removed from a world for any reason (including a chunk unloading).
+ * Note: The entity is updated prior to this event being called, as such, the entity's world may not be equal to {@link #getWorld()}.
+ */
+public class EntityRemoveFromWorldEvent extends EntityEvent {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+
+    @NotNull private final World world;
+
+    @NotNull private final RemovalReason reason;
+
+    @ApiStatus.Internal
+    public EntityRemoveFromWorldEvent(@NotNull Entity entity, @NotNull World world, @NotNull RemovalReason reason) {
+        super(entity);
+        this.world = world;
+        this.reason = reason;
+    }
+
+    /**
+     * @return The world that the entity is being removed from
+     */
+    @NotNull
+    public World getWorld() {
+        return this.world;
+    }
+
+    @NotNull
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    public RemovalReason getReason() {
+        return reason;
+    }
+
+    /**
+     * Represents various ways an entity gets removed.
+     */
+    public enum RemovalReason {
+        /**
+         * When an entity dies.
+         */
+        DEATH,
+        /**
+         * When an entity does despawn. This includes mobs which are too far away,
+         * items or arrows which lay to long on the ground or area effect cloud.
+         */
+        DESPAWN,
+        /**
+         * When an entity gets removed because it drops as an item.
+         * For example, trident or falling sand.
+         * <p>
+         * <b>Note:</b> Depending on other factors, such as gamerules, no item will actually drop,
+         * the cause, however, will still be drop.
+         */
+        DROP,
+        /**
+         * When an entity gets removed because it enters a block.
+         * For example, bees or silverfish.
+         */
+        ENTER_BLOCK,
+        /**
+         * When an entity gets removed because it exploded.
+         * For example, creepers, tnt or firework.
+         */
+        EXPLODE,
+        /**
+         * When an entity gets removed because it hit something. This mainly applies to projectiles.
+         */
+        HIT,
+        /**
+         * When an entity gets removed because it merges with another one.
+         * For example, items or xp.
+         */
+        MERGE,
+        /**
+         * When an entity gets removed because it is too far below the world.
+         * This only applies to entities which get removed immediately,
+         * some entities get damage instead.
+         */
+        OUT_OF_WORLD,
+        /**
+         * When an entity gets removed because it got pickup.
+         * For example, items, arrows, xp or parrots which get on a player shoulder.
+         */
+        PICKUP,
+        /**
+         * When an entity gets removed with a player because the player quits the game.
+         * For example, a boat which gets removed with the player when he quits.
+         */
+        PLAYER_QUIT,
+        /**
+         * When a plugin manually removes an entity.
+         */
+        PLUGIN,
+        /**
+         * When an entity gets removed because it transforms into another one.
+         */
+        TRANSFORMATION,
+        /**
+         * When the chunk an entity is in gets unloaded.
+         */
+        UNLOAD,
+        DISCARD;
+
+        public static RemovalReason fromBukkitCause(org.bukkit.event.entity.EntityRemoveEvent.Cause cause) {
+            return java.util.Arrays.stream(values()).filter(removalReason -> removalReason.name().equalsIgnoreCase(cause.name())).findFirst().orElseThrow(() -> new IllegalArgumentException("Can not convert Cause " + cause + " to RemovalReason"));
+        }
+    }
+}
diff --git a/src/main/java/org/bukkit/event/entity/EntityRemoveEvent.java b/src/main/java/org/bukkit/event/entity/EntityRemoveEvent.java
index e32df91d911bae42c8137c6f952a6ac6a94d27e0..8ed5d1ccc44951089999db360219b556db89b4ba 100644
--- a/src/main/java/org/bukkit/event/entity/EntityRemoveEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityRemoveEvent.java
@@ -1,5 +1,6 @@
 package org.bukkit.event.entity;
 
+import com.destroystokyo.paper.event.entity.EntityRemoveFromWorldEvent;
 import org.bukkit.entity.Entity;
 import org.bukkit.event.HandlerList;
 import org.jetbrains.annotations.ApiStatus;
@@ -11,8 +12,9 @@ import org.jetbrains.annotations.NotNull;
  * This event should only be used for monitoring. The result
  * of modifying the entity during or after this event is unspecified.
  * This event is not called for a {@link org.bukkit.entity.Player}.
+ * @deprecated use {@link EntityRemoveFromWorldEvent} instead
  */
-@ApiStatus.Experimental
+@Deprecated(forRemoval = true)
 public class EntityRemoveEvent extends EntityEvent {
 
     private static final HandlerList handlers = new HandlerList();
@@ -112,5 +114,6 @@ public class EntityRemoveEvent extends EntityEvent {
          * When the chunk an entity is in gets unloaded.
          */
         UNLOAD,
+        DISCARD
     }
 }
